### ISSUE TITLE
Implement a DataLoaderDSL for batching

### ISSUE THREAD

**Issue by jeggy** [aPureBase/KGraphQL#19]
Implement a [DataLoader ](https://github.com/facebook/dataloader) style batching feature.

I've been thinking about a structure like this:

```kotlin
...
data class Tree(val id: String, val parentId: String?)
...
type<Tree> {
    // String - defines the key that will be sent from the [prepare] into [loader]
    // List<Tree> - defines the return type that the [loader] is required to return.
    // the loader is then required to return it in a map format like Map<String, List<Tree>>
    dataProperty<String, List<Tree>>("children") {
        // Step 2: This will only be called once.
        loader { keys: List<String> ->
            keys.map{ id -> id to listOf(Tree("SomeId", id)) }.toMap()
        }
    
        // Step 1: This will be called for each node in the list, or only once if it's not a list
        prepare { parent: Tree -> parent.id }
    }
}
```

_Anyone is welcome with some input on how this could be achieved._

**Comment by Leopard2A5** [comment]
+1 this would really help!

**Comment by nidomiro** [comment]
@jeggy I'm currently developing a library for this purpose in pure Kotlin: https://github.com/nidomiro/KDataLoader

**Comment by jeggy** [comment]
@nidomiro This looks great. I'll try to sketch out a new `RequestExector` that will take use of your library 😃 

I'm thinking about making a single coroutineScope per level of nested querying and then after all the coroutineScopes have been created, we could call the `dispatch` on each level that's needed. 
So for each field that does have a dataloader that field execution would just suspend until all the scopes have been defined, but the fields not having a dataloader would just start immediately.

**Comment by nidomiro** [comment]
If I understood you right, you are thinking of a "Breadth-first search"-like Execution. And after finishing one depth-Level calling `dispatch`.
This sounds like a good solution.

Does it help you if I create a DataLoaderFactory? Then one can configure the loader once through a DSL and KGraphQL creates a new Dataloader for every `schema.execute()` with this factory.

**Comment by jeggy** [comment]
A DataLoaderFactory would be perfect 👍

**Comment by jeggy** [comment]
I'm struggling to find a solution to have an overview of coroutines structure, so it can be controlled so we can call the dispatch at the correct time.
I currently just did this hack: https://github.com/aPureBase/KGraphQL/pull/69/files#diff-857d4482b068cd916df35079d8920e6aR198-R223

If you have some idea on how to structure the execution of this, any help would be appreciated.

**Comment by nidomiro** [comment]
Yea, this is a really tricky one...
Currently I have no proper solution.
Every "solution" I come up with will reduce the parallelism in some way.

Maybe counting how many `loadAsync` calls are made for one loader and comparing them to how many should be made is a way. => For example: This level requests `n` objects of type `x`, so if the `loadAsync` was called `n` times for type `x` (in this level) then call `dispatch`.

I will implement some statistics right into the DataLoader.
And I have to verify that calling `dispatch` in parallel will not start the dispatch-process multiple times for the same entries.

**Comment by jeggy** [comment]
We could maybe just drop the parallelism for now. Having support for this will bring way better performance than having parallelism.

**Comment by nidomiro** [comment]
> We could maybe just drop the parallelism for now. Having support for this will bring way better performance than having parallelism.

Yes, we can add it later again.
It is not helpful if it stops us from implementing more features.
This would also help with reimplementing Subscriptions ;)
But we have to keep in mind, that we want to implement it again in the future.


I just published KDataLoader 0.1.0, which includes statistics and the factory.
Factory examples: https://github.com/nidomiro/KDataLoader/blob/5e5e1f53ffd8476339eaa221a718bef8a6a23381/src/jvmTest/kotlin/nidomiro/kdataloader/DSLTest.kt#L102-160

Statistics examles: https://github.com/nidomiro/KDataLoader/blob/5e5e1f53ffd8476339eaa221a718bef8a6a23381/src/jvmTest/kotlin/nidomiro/kdataloader/DataLoaderStatisticsTest.kt

**Comment by jeggy** [comment]
I got it a working sample now 🤩 

I want to cleanup the code and get more unit tests to hit more edge cases before merging over to master.

Sample usage:
![image](https://user-images.githubusercontent.com/51459/72535263-28c96000-3870-11ea-97cd-799311597489.png)

I still have some issues with nested queries using the same dataLoader, like this test: https://github.com/aPureBase/KGraphQL/blob/7469768b8c4d4354d1103190cf892ca2ef36db47/kgraphql/src/test/kotlin/com/apurebase/kgraphql/DataLoaderTest.kt#L350-L367

I couldn't have done it without help from the guys at [#coroutines kotlin slack channel: https://kotlinlang.slack.com/archives/C1CFAFJSK/p1579026456122300](https://kotlinlang.slack.com/archives/C1CFAFJSK/p1579026456122300)

**Comment by FabioGimmillaro** [comment]
Really looking forward to this feature :)

**Comment by jeggy** [comment]
I have created a new release `0.10.0` that has this 😃 

I haven't created any documentation for this yet, so for now I'll just post sample usage here:
```kotlin
data class Value(val id: Int, val name: String)

val parent1 = Value(1, "Parent 1")
val parent2 = Value(2, "Parent 2")
val child1 = Value(3, "Child 1")
val child2 = Value(4, "Child 2")
val child3 = Value(5, "Child 3")

val source = mapOf(
    1 to listOf(child1, child2),
    2 to listOf(child3)
)

fun main() {
    val schema = KGraphQL.schema {
        configure {
            executor = Executor.DataLoaderPrepared
        }

        query("parents") {
            resolver { -> listOf(parent1, parent2) }
        }

        type<Value> {
            dataProperty<Int, List<Value>>("children") {
                prepare { parent -> parent.id }
                loader { parentIds ->
                    parentIds.map { ExecutionResult.Success(source.getValue(it)) }
                }
            }
        }
    }

    schema.executeBlocking("""
        {
            parents {
                id
                name
                children {
                    id
                    name
                }
            }
        }
    """).let(::println)
}
```

Using this new executor currently has some problems, which in most cases shouldn't be a problem.
* useDefaultPrettyPrint doesn't work
* As this implementation is spinning up a lot of coroutines, the order of fields returned aren't guaranteed to match the one that was requested. Meaning a request like this: 
```graphql
{
  parents {
    name
    children { name }
  }
}
```

Could return in a response like this:
```json
{
  "data": {
    "parents": [
      {
        "children": null,       # Children should come after "name".
        "name": "Parent 1"
      }
    ]
  }
}
```

Other than please try it out and any feedback is welcome 😃

**Comment by FabioGimmillaro** [comment]
Works like a charm 
Thank you :D

**Comment by jeggy** [comment]
This has been added to the documentation here: [Data Loaded Properties](https://kgraphql.io/Reference/Type%20System/objects-and-interfaces/#data-loaded-properties)

If anyone has any more suggestions on improving this or finds any issues please open a new issue
